Reference/Routing/custom-controllers.md

```diff
@@ -2,167 +2,246 @@
 
+The default inherits statement:
 ```csharp
 @inherits Umbraco.Web.Mvc.UmbracoViewPage
 ```
 or if you are using modelsbuilder:
 
 ```csharp
-@inherits Umbraco.Web.Mvc.UmbracoViewPage<HomePage>
+@inherits Umbraco.Web.Mvc.UmbracoViewPage<ProductPage>
 ```
-(A model is generated for each document type to give strongly typed access to properties in the template)
+(where <> contains a model generated for each document type to give strongly typed access to the document type properties in the template view)
+
+To use a specific custom view model, the @inherits directive will need to be updated to reference your custom model using the `Umbraco.Web.Mvc.UmbracoViewPage<T>` format where 'T' is the type of your custom model.
 
-If you are returning a custom model, then this directive will need to change because your custom model will not be an instance of `Umbraco.Web.Models.ContentModel`. Instead change your @inherits directive to inherit from `Umbraco.Web.Mvc.UmbracoViewPage<T>` where 'T' is the type of your custom model. So for example, if your custom model is of type 'MyCustomModel' then your @inherits directive will look like:
+So for example, if your custom model is of type 'MyProductModel' then your @inherits directive will look like:
 
 ```csharp
-@inherits Umbraco.Web.Mvc.UmbracoViewPage<MyCustomModel>
+@inherits Umbraco.Web.Mvc.UmbracoViewPage<MyProductModel>
 ```
 
-Please note that if your template uses a layout that expects the model to be of type `Umbraco.Web.Models.ContentModel` then changing the template to inherit from `Umbraco.Web.Mvc.UmbracoViewPage<MyCustomModel>` will cause an exception. This is due to the way ASP.NET MVC works with strongly typed views: the requirement for a specific type applies all the way from the top-most layout down to the template. There are two ways to solve this problem:
+:::note
+Views in MVC will likely specify a master view to use as the common layout for the site html, when using a custom view model it's necessary to make sure this doesn't conflict with any implementation in the master layout view.
+eg: if your master layout view is inheriting from a specific model `UmbracoViewPage<SpecificModel>` and using a property from SpecificModel that isn't available in your custom model an exception will be thrown!
+To avoid this you could:
+_ Keep your Master layout view 'generically typed', eg only have @inherits UmbracoViewPage, and use Model.Value syntax to access properties.
+or
+_ Break the dependency on `Umbraco.Web.Models.ContentModel` in your master layout by having it instead inherit from `Umbraco.Web.Mvc.UmbracoViewPage<ISomeInterface>` (where ISomeInterface is implemented by all your models and contains the properties that the master layout view uses).
+or
+_ Ensure your custom models inherit from whichever class is used to strongly type the master layout view.
```

Comment:
```suggestion
* Keep your Master layout view 'generically typed', eg. only have `@inherits UmbracoViewPage`, and use Model.Value syntax to access properties.
or
* Break the dependency on `Umbraco.Web.Models.ContentModel` in your master layout by having it instead inherit from `Umbraco.Web.Mvc.UmbracoViewPage<ISomeInterface>`. This would be where ISomeInterface is implemented by all your models and contains the properties that the master layout view uses.
or
* Ensure your custom models inherit from whichever class is used to strongly type the master layout view.
```